servers/src/main/java/tachyon/worker/netty/DataServerHandler.java

```diff
@@ -126,6 +137,52 @@ private void handleBlockRequest(final ChannelHandlerContext ctx, final RPCBlockR
     }
   }
 
+  // TODO: This write request handler is very simple in order to be stateless. Therefore, the block
+  // file is opened and closed for every request. If this is too slow, then this handler should be
+  // optimized to keep state.
+  private void handleBlockWriteRequest(final ChannelHandlerContext ctx,
+      final RPCBlockWriteRequest req) throws IOException {
+    final long userId = req.getUserId();
+    final long blockId = req.getBlockId();
+    final long offset = req.getOffset();
+    final long length = req.getLength();
+    final DataBuffer data = req.getPayloadDataBuffer();
+
+    BlockWriter writer = null;
+    try {
+      req.validate();
+      ByteBuffer buffer = data.getReadOnlyByteBuffer();
+      if (buffer.remaining() <= 0) {
+        throw new IOException("Empty buffer to write.");
+      }
+
+      if (offset == 0) {
```

Comment:
It is only slightly less complicated, since space allocation is different in each case.